go/mounter/keybase-mount-helper/main.go

```diff
@@ -0,0 +1,194 @@
+		"path to an existing directory that will serve as the true mountpoint "+
+			"for the Keybase user")
+}
+
+func checkLink(firstMountLink, originalFirstMountTarget, userMount string) (
+	firstMount bool, makeMount bool, err error) {
+	currLink, err := os.Readlink(firstMountLink)
+	if err != nil {
+		return false, false, err
+	}
+
+	if currLink == userMount {
+		// User already owns the link.
+		return true, false, nil
+	}
+
+	if currLink != originalFirstMountTarget {
+		// Another user already owns this link.
+		return false, false, nil
+	}
+
+	// User may try to make the link.
+	return true, true, nil
+}
+
+func takeLock(firstMountLink string) (lockFile *os.File, err error) {
+	// Take an flock to ensure only one user tries to get the link.
+	lockFilename := firstMountLink + ".lock"
+	lockFile, err = os.Create(lockFilename)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		if err != nil {
+			lockErr := lockFile.Close()
+			if lockErr != nil {
+				fmt.Fprintf(os.Stderr, "Couldn't close lock file: %+v", lockErr)
+			}
+		}
+	}()
+
+	err = syscall.Flock(int(lockFile.Fd()), syscall.LOCK_EX)
+	if err != nil {
+		return nil, err
+	}
+	return lockFile, nil
+}
+
+func checkAndSwitchMount(
+	firstMountLink, originalFirstMountTarget, userMount string) (
+	firstMount bool, err error) {
+	firstMount, makeMount, err := checkLink(
+		firstMountLink, originalFirstMountTarget, userMount)
+	if err != nil {
+		return false, err
+	}
+	if !makeMount {
+		return firstMount, nil
+	}
+
+	lockFile, err := takeLock(firstMountLink)
+	if err != nil {
+		return false, err
+	}
+
+	defer func() {
+		lockErr := lockFile.Close()
+		if err == nil {
+			err = lockErr
+		}
+	}()
+
+	// Make sure the link is in the same state after taking the lock.
+	firstMount, makeMount, err = checkLink(
+		firstMountLink, originalFirstMountTarget, userMount)
+	if err != nil {
+		return false, err
+	}
+	if !makeMount {
+		return firstMount, nil
+	}
+
+	// Make a symlink for the new user, and swap it over the existing
+	// symlink.
+	tmpLink := firstMountLink + ".tmp"
+	err = os.Symlink(userMount, tmpLink)
+	if err != nil {
+		return false, err
+	}
+
+	defer func() {
+		if err != nil {
+			rmErr := os.Remove(tmpLink)
+			if rmErr != nil {
+				fmt.Fprintf(os.Stderr,
+					"Can't remove %s on error: %+v\n", tmpLink, rmErr)
+			}
+		}
+	}()
+
+	err = os.Rename(tmpLink, firstMountLink)
+	if err != nil {
+		return false, err
+	}
+	return true, nil
+}
+
+func main() {
+	flag.Parse()
+
+	// Figure out the helper UID.
+	khUser, err := user.Lookup(kbUsername)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Cannot get %s user: %+v\n", kbUsername, err)
+		os.Exit(1)
+	}
+	khUID, err := strconv.Atoi(khUser.Uid)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Bad UID for %s user: %+v\n", kbUsername, err)
+		os.Exit(1)
+	}
+
+	// Lock the thread (because SYS_SETUID only sets the UID of the
+	// current OS thread) and switch to the helper user.
+	runtime.LockOSThread()
```

Comment:
Wow how did you think of this! Impressive.

Reply:
Stumbled across this while looking up setuid stuff: https://github.com/golang/go/issues/1435